### PR TITLE
Prevent exception when creating a Divider borderSide

### LINKED ISSUE #39533: ListTile.divideTiles now requires context: The method 'inheritFromWidgetOfExactType' was called on null.
Upgraded to the latest Flutter and it looks like there is an error with ListTiles and/or ListView

I/flutter ( 8416): ══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
I/flutter ( 8416): The following NoSuchMethodError was thrown building FormTileGroup(dirty):
I/flutter ( 8416): The method 'inheritFromWidgetOfExactType' was called on null.
I/flutter ( 8416): Receiver: null
I/flutter ( 8416): Tried calling: inheritFromWidgetOfExactType(DividerTheme)
I/flutter ( 8416): 
I/flutter ( 8416): User-created ancestor of the error-causing widget was:
I/flutter ( 8416):   ListView
I/flutter ( 8416):   file:///Users/user/AndroidStudioProjects/conquered_time/lib/src/drawers/profile_drawer.dart:279:20
I/flutter ( 8416): 
I/flutter ( 8416): When the exception was thrown, this was the stack:
I/flutter ( 8416): #0      Object.noSuchMethod (dart:core-patch/object_patch.dart:51:5)
I/flutter ( 8416): #1      DividerTheme.of (package:flutter/src/material/divider_theme.dart:162:47)
I/flutter ( 8416): #2      Divider.createBorderSide (package:flutter/src/material/divider.dart:122:36)
I/flutter ( 8416): #3      ListTile.divideTiles (package:flutter/src/material/list_tile.dart:740:25)
I/flutter ( 8416): #4      _SyncIterator.moveNext (dart:core-patch/core_patch.dart:144:12)
I/flutter ( 8416): #5      new List.from (dart:core-patch/array_patch.dart:45:19)
I/flutter ( 8416): #6      Iterable.toList (dart:core/iterable.dart:398:5)
I/flutter ( 8416): #7      FormTileGroup.build (package:conquered_time/src/ui/controls/form_controls.dart:31:38)

### PR DESCRIPTION
**Pull request by rami-a**
## Description

Check for a null context in the `Divider.createBorderSide` function and ensure no exception occurs.

## Related Issues

Fixes #39533 

## Tests

I added the following tests:

* A test that would have thrown exceptions before this change

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Review by willlarche — APPROVED** [review]

File: packages/flutter/test/material/divider_test.dart [hunk 1]
```diff
@@ -203,4 +203,10 @@ void main() {
     expect(lineRect.top, dividerRect.top + customIndent);
     expect(lineRect.bottom, dividerRect.bottom - customIndent);
   });
+
+  // Regression test for https://github.com/flutter/flutter/issues/39533
+  testWidgets('createBorderSide does not throw exception with null context', (WidgetTester tester) async {
+    Divider.createBorderSide(null);
```
Comment (johnsonmh) [inline packages/flutter/test/material/divider_test.dart]: consider using `isNot(throwsAssertionError)`

```suggestion
expect(() => Divider.createBorderSide(null), isNot(throwsAssertionError));
```

File: packages/flutter/lib/src/material/divider.dart [hunk 2]
```diff
@@ -117,9 +117,20 @@ class Divider extends StatelessWidget {
   /// ```
   /// {@end-tool}
   static BorderSide createBorderSide(BuildContext context, { Color color, double width }) {
+    final Color themeColor = context != null ? (DividerTheme.of(context).color ?? Theme.of(context).dividerColor) : null;
+    final double themeWidth = context != null ? DividerTheme.of(context).thickness : null;
+    final double resolvedWidth = width ?? themeWidth ?? 0.0;
```
Comment (johnsonmh) [inline packages/flutter/lib/src/material/divider.dart]: I think the typical naming for something like this is `effectiveWidth`

File: packages/flutter/lib/src/material/divider.dart [hunk 3]
```diff
@@ -117,9 +117,20 @@ class Divider extends StatelessWidget {
...
+
```
Comment (johnsonmh) [inline packages/flutter/lib/src/material/divider.dart]: consider adding 

`final Color effectiveColor = color ?? themeColor'`

Then below `color == null && themeColor == null` could be `effectiveColor == null`

**Review by johnsonmh — APPROVED** [review]
LGTM

File: packages/flutter/test/material/divider_test.dart [hunk 4]
```diff
@@ -203,4 +203,10 @@ void main() {
...
+    // Passing a null context used to throw an exception but no longer does.
```
Comment (clocksmith) [inline packages/flutter/test/material/divider_test.dart]: // No expect needed here because passing null does not throw an exception

**Review by clocksmith — APPROVED** [review]

File: packages/flutter/test/material/divider_test.dart (hunk 1 above)
Comment (rami-a) [reply packages/flutter/test/material/divider_test.dart]: done

File: packages/flutter/lib/src/material/divider.dart (hunk 2 above)
Comment (rami-a) [reply packages/flutter/lib/src/material/divider.dart]: Done

File: packages/flutter/lib/src/material/divider.dart (hunk 3 above)
Comment (rami-a) [reply packages/flutter/lib/src/material/divider.dart]: Done

File: packages/flutter/test/material/divider_test.dart (hunk 4 above)
Comment (rami-a) [reply packages/flutter/test/material/divider_test.dart]: Added an expect instead

File: packages/flutter/lib/src/material/divider.dart [hunk 5]
```diff
@@ -117,9 +117,21 @@ class Divider extends StatelessWidget {
   /// ```
   /// {@end-tool}
   static BorderSide createBorderSide(BuildContext context, { Color color, double width }) {
+    final Color themeColor = context != null ? (DividerTheme.of(context).color ?? Theme.of(context).dividerColor) : null;
+    final Color effectiveColor = color ?? themeColor;
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/divider.dart]: I think this should be 
```
final Color effectiveColor = color ?? (context != null ? DividerTheme.of(context).color : null);
```
DividerTheme.of() already defers to Theme.of() and if color is specified there's no need to create an (inherited widget) dependency on a theme.

File: packages/flutter/lib/src/material/divider.dart [hunk 6]
```diff
@@ -117,9 +117,21 @@ class Divider extends StatelessWidget {
...
+    final double themeWidth = context != null ? DividerTheme.of(context).thickness : null;
+    final double effectiveWidth = width ?? themeWidth ?? 0.0;
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/divider.dart]: Similar comment here:
```
final double effectiveWidth =  width ?? (context != null ? DividerTheme.of(context).thickness : null) ?? 0.0;
```

File: packages/flutter/lib/src/material/divider.dart [hunk 7]
```diff
@@ -117,9 +117,21 @@ class Divider extends StatelessWidget {
   /// ```
   /// {@end-tool}
   static BorderSide createBorderSide(BuildContext context, { Color color, double width }) {
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/divider.dart]: We should document what happens if BuildContext is null. This is a bit of an odd method; usually this kind of thing would require a non-null context or else context would be a named parameter.

**Review by HansMuller — APPROVED** [review]
LGTM with some tweaks.

File: packages/flutter/lib/src/material/divider.dart (hunk 5 above)
Comment (rami-a) [reply packages/flutter/lib/src/material/divider.dart]: Unfortunately, this wouldn't allow us to use Theme.of().dividerColor as a fallback (the legacy behavior), so I'd have to keep that

Comment (HansMuller) [reply packages/flutter/lib/src/material/divider.dart]: Good point. So long as it's all part of an expression that short-circuits if color is non-null.

File: packages/flutter/lib/src/material/divider.dart (hunk 6 above)
Comment (rami-a) [reply packages/flutter/lib/src/material/divider.dart]: done

File: packages/flutter/lib/src/material/divider.dart (hunk 7 above)
Comment (rami-a) [reply packages/flutter/lib/src/material/divider.dart]: done

File: packages/flutter/lib/src/material/divider.dart (hunk 5 above)
Comment (rami-a) [reply packages/flutter/lib/src/material/divider.dart]: but I consolidated it into one variable

Comment (HansMuller) [reply packages/flutter/lib/src/material/divider.dart]: Thanks, looks good.

**Pull request merged**